### ISSUE TITLE
[Performance] GSpMM regression with FP16

### ISSUE THREAD

**Issue by yaox12** [dmlc/dgl#4270]
I spent some time profiling the GAT example with AMP in https://docs.dgl.ai/en/0.9.x/guide/mixed_precision.html and want to know why we didn't obtain performance gain from FP16. I observed regression in both the forward and backward phases.

## Performance on A100

|      | AMP | FP32 |
|--------------|--------|--------|
| forward  |  254 ms | 160 ms  |
| backward |  482 ms | 369 ms |
| step     |  738 ms | 529 ms |

Take the forward phase for example, there are three `EdgeSoftmax-GSpMM` pairs.
- AMP 
![image](https://user-images.githubusercontent.com/3831900/179916055-e7c714f9-60ab-4577-88c5-b26e25c4d7b2.png)
- FP32
![image](https://user-images.githubusercontent.com/3831900/179916242-38875886-859d-4649-b62a-dea5a756f104.png)

Elapsed times (ms) of the `EdgeSoftmax-GSpMM` pairs x3.
|              | AMP    | FP32   |
|--------------|--------|--------|
| EdgeSoftmax1 | 20.896 | 22.055 |
| GSpMM1        | 73.851 | 31.890 |
| EdgeSoftmax2 | 20.865 | 22.080 |
| GSpMM2        | 73.797 | 31.882 |
| EdgeSoftmax3 | 20.869 | 22.154 |
| GSpMM3        | 27.528 | 13.422 |

We can see that the `EdgeSoftmax` is a bit faster with AMP while it's `GSpMM` that results in the regression.

Dig it a bit deeper, `GSpMM` with FP32 will go through the `cusparse::csrmm` code path while adopting `dgl::aten::cuda::SpMMCsrKernel` with FP16 inputs. I'll look into the performance issue on cusparse with FP16.

### ENV
- CUDA_VERSION: 11.7.0.022
- CUDA_DRIVER_VERSION: 515.43.04
- CUSPARSE_VERSION: 11.7.3.50
- GPU: NVIDIA A100 80GB PCIe
- PyTorch: 1.12.0a0+8a1a93a

**Comment by yzh119** [comment]
I encountered a similar issue a year ao, I remember that cusparse fp16 requires some alignment (e.g. the array pointer address should be multiple of 16/32/64, etc) to be efficient.

**Comment by nv-dlasalle** [comment]
@yaox12 Can you add cuda versions?

**Comment by yaox12** [comment]
> @yaox12 Can you add cuda versions?

Added.

**Comment by yaox12** [comment]
I have updated some numbers in V100. It seems that the `cusparseSpMM` FP16 regression happens only on A100. 

On V100, if we replace the DGL's custom SpMM F16 kernel with `cusparseSpMM` FP16, we can achieve ~15% end2end speedup, and ~10% end2end speedup over `cusparseSpMM` FP32.

**Comment by yzh119** [comment]
@yaox12 I just found a code snippet about the alignment issue written by @nv-dlasalle in the last year:
https://github.com/nv-dlasalle/dgl/commit/5fc6e9bfc5fbd59e0cf5dbc4510883a5a124a467

The matrix column needs to be aligned to 128 bytes for best performance.

**Comment by yaox12** [comment]
I find something interesting. First, I don't think this is an issue related to data alignment. Because in this GAT model, the node features are projected from 602 to 256-dimension tensors before invoking GSpMM.

I randomly generated a graph with the same number of nodes and edges as Reddit via `dgl.rand_graph`. Then I benchmark `cusparseSpMM` on them and get the following numbers (in ms):
<table>
<thead>
  <tr>
    <th></th>
    <th colspan="2">feat_size = 256</th>
    <th colspan="2">feat_size = 602</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td></td>
    <td>random graph</td>
    <td>Reddit</td>
    <td>random graph</td>
    <td>Reddit</td>
  </tr>
  <tr>
    <td>FP16</td>
    <td>93.76</td>
    <td>97.30</td>
    <td>239.44</td>
    <td>257.41</td>
  </tr>
  <tr>
    <td>FP32</td>
    <td>80.73</td>
    <td>49.35</td>
    <td>184.64</td>
    <td>103.18</td>
  </tr>
</tbody>
</table>

It seems that `cusparseSpMM` with FP32 has some special optimization for certain topologies/degree distributions while FP16 doesn't.

Here is my code:
```python
import time
import torch
import dgl
from dgl.data import RedditDataset
from dgl.ops import gspmm

# g = dgl.rand_graph(232965, 114615892)
# dgl.save_graphs('g.bin', g)

g = RedditDataset(raw_dir='.')[0]
# g = dgl.load_graphs('g.bin')[0][0]

dtype = torch.float16
device = 'cuda:0'
g = g.int().to(device)

def run(shp, msg, reducer="sum"):

    u = torch.rand((g.number_of_src_nodes(), shp[0]), dtype=dtype, device=device)
    e = torch.rand((g.number_of_edges(), shp[1]), dtype=dtype, device=device)

    tic = time.time()
    for _ in range(10):
        gspmm(g, msg, reducer, u, e)
    print(f'{shp[0]}, {dtype}: {(time.time() - tic) * 100:.2f} ms')

# run((256, 1), 'mul')
run((602, 1), 'mul')
```

**Comment by yzh119** [comment]
Not only about feature size you should also check the pointer address of each operand (A, B and C).
To the best of my knowledge, achieving such performance on Reddit does not require any special optimizations.

**Comment by yzh119** [comment]
Sputnik also has some issues about alignment for fp16: https://github.com/facebookresearch/xformers/issues/15

**Comment by chang-l** [comment]
@yaox12 
Thank you for sharing your code. I tested it on my side (A100, cuda11.7), but I observed different results: 
| feat dim | fp32 (ms) | fp16 (ms) | Speed-up of fp16
| ------------- | -------| -------|-------
| 32 | 17.6 | 18.75 | 0.938666667
| 64 | 21.35 | 22.1 | 0.966063348
| 128 | 40.46 | 28.7 | 1.409756098
| 256 | 78.24 | 55.59 | 1.407447383
| 512 | 156.18 | 100.48 | 1.554339172
| 602 | 203.78 | 140.86 | 1.446684651

I also forced the code using cusparseSpMM for FP16. Note that I used following mixed-precision formats when using cusparseSpMM for fp16:

| A/B	| C	| computeType
| ----------- | -----------| -----------
CUDA_R_16F |	CUDA_R_16F	| CUDA_R_32F

**Comment by frozenbugs** [comment]
Hi @yaox12 , is this done already?

**Comment by yaox12** [comment]
Still tracking it internally.
